src/main/java/net/neoforged/neoforge/transfer/handlers/templates/resources/StacksResourceHandler.java

```diff
@@ -0,0 +1,287 @@
+
+    protected NonNullList<S> stacks;
+    protected final S emptyStack;
+    protected final Codec<NonNullList<S>> codec;
+
+    private final List<StackJournal> snapshotJournals;
+
+    protected StacksResourceHandler(int size, S emptyStack, Codec<S> stackCodec) {
+        this(NonNullList.withSize(size, emptyStack), emptyStack, stackCodec);
+    }
+
+    protected StacksResourceHandler(NonNullList<S> stacks, S emptyStack, Codec<S> stackCodec) {
+        this.stacks = mutableCopyOf(stacks);
+        this.emptyStack = emptyStack;
+        // Don't use NonNullList.codecOf because it creates an unmodifiable list
+        this.codec = stackCodec.listOf().xmap(StacksResourceHandler::mutableCopyOf, Function.identity());
+        this.snapshotJournals = new ArrayList<>(this.stacks.size());
+        for (int i = 0; i < this.stacks.size(); i++) {
+            snapshotJournals.add(new StackJournal(i));
+        }
+    }
+
+    private static <T> NonNullList<T> mutableCopyOf(Collection<T> list) {
+        NonNullList<T> ret = NonNullList.createWithCapacity(list.size());
+        ret.addAll(list);
+        return ret;
+    }
+
+    @Override
+    public void serialize(ValueOutput output) {
+        output.store(VALUE_IO_KEY, codec, stacks);
+    }
+
+    @Override
+    public void deserialize(ValueInput input) {
+        Optional<NonNullList<S>> optional = input.read(VALUE_IO_KEY, codec);
+        if (optional.isEmpty()) return;
+
+        stacks = optional.get();
+    }
+
+    /**
+     * Directly overwrites the contents of the handler.
+     *
+     * <p>Note that this method can be used as an {@link IndexModifier}, for usage in {@link ResourceHandlerSlot}.
+     *
+     * @param index    index to change
+     * @param resource new resource at the index
+     * @param amount   new amount at the index
+     * @throws IllegalArgumentException if either the amount is negative; or if the resource is non-empty for a 0 amount
+     */
+    public void set(int index, T resource, int amount) {
+        TransferPreconditions.checkNonNegative(amount);
+        if (resource.isEmpty() && amount > 0) {
+            throw new IllegalArgumentException("Resource is empty but the amount is positive: " + amount);
+        }
+
+        S oldContents = stacks.set(index, getStackFrom(resource, amount));
+        onContentsChanged(index, oldContents);
+    }
+
+    /**
+     * Retrieves the resource from a stack.
+     * In the case of an {@link ItemStack} an {@link ItemResource} would be returned for example.
+     */
+    protected abstract T getResourceFrom(S stack);
+
+    /**
+     * Retrieves the amount from a stack.
+     * In the case of an {@link ItemStack} {@linkplain ItemStack#getCount() the count} would be returned for example.
+     */
+    protected abstract int getAmountFrom(S stack);
+
+    /**
+     * Creates a stack from a resource and an amount.
+     */
+    protected abstract S getStackFrom(T resource, int amount);
```

Comment:
Do we want impls of this to return `emptyStack` if the resource is empty or the amount is zero? Rather than creating a new empty stack instance when we already have one available?